src/Castle.Core.AsyncInterceptor/Invocations/InvocationBase.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Copyright (c) 2016 James Skimming. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See LICENSE in the project root for license information.
+
+namespace Castle.DynamicProxy.Invocations
+{
+    using System;
+    using System.Reflection;
+    using System.Runtime.CompilerServices;
+
+    /// <inheritdoc />
+    internal abstract class InvocationBase : IAsyncInvocation
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="InvocationBase"/> class.
+        /// </summary>
+        /// <param name="invocation">
+        /// The invocation.
+        /// </param>
+        protected InvocationBase(IInvocation invocation)
+        {
+            Invocation = invocation;
```

Comment:
Don't you want to implement the Proceed method as well?
Here I see the mix of IInvocation and IAsyncInvocation methods

Reply:
Good question. 

I have intentionally not added `Proceed` and `ReturnValue` to `IAsyncInvocation` as it should not be visible to the Interceptor.

The return-value handling and proceed is implemented separately in each pair of
```csharp 
class FunctionInvocation<TResult> : InvocationBase, IFunctionInvocation<TResult>
```
The `InvocationBase` class basically wraps `IInvocation` from Castle.